torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,536 @@
+// 4) Identity: there exist two different elements 0 and 1 such that
+//    a + 0 = a and a * 1 = a.
+// 5) Additive inverses: for every a, there exists an element, denoted -a, such
+//    that a + (-a) = 0.
+// 6) Defition of subtraction: a - b := a + (-b)
+// 7) Multiplicative inverses: for every a != 0, there exists an element,
+//    denoted by 1/a, such that a * (1/a) = 1.
+// 8) Defition of division: a / b := a * (1/b)
+//
+// Thanks to the field properties, div of real number has the following
+// properties:
+// 9) Associativity:
+// - a*(b/c) = (a*b)/c
+// - a/(b*c) = (a/b)/c
+// - a/(b/c) = (a/b)*c = (a*c)/b
+// 10) Right Distributivity: (a+b)/c = a/c + b/c
+//
+// For the set of integers Z, 7) is no longer true, therefore, the division can
+// not be defined as 8), and as a result, neither 9) nor 10) is true anymore.
+// So, the mathematical structure of integer arithmetics is not as strong as a
+// field.
+//
+// The integer arithmetics form an Euclidean domain:
+// https://en.wikipedia.org/wiki/Euclidean_domain
+//
+// The relationship between Euclidean domains and other algebraic structures is
+// as follows (A > B means A is more general than B, that is, every B is an A,
+// that is, B has more structure/stronger properties than A):
+//   rings > commutative rings > [...] > Euclidean domains > fields
+//
+// where "commutative rings" is almost a field except that:
+// - in 4), 0 and 1 are not required to be different
+// - in 7) and 8) are not required to be true
+//
+// the [...] above can be further expanded as:
+//   > integral domains > GCD domains > unique factorization domains >
+//
+// The expanded [...] and Euclidean domains are made by gradually adding basic
+// properties of integers to the recipe. These properties include:
+// - integral domains:
+//   - 0 != 1
+//   - if a != 0 and b != 0, then ab != 0
+//   - cancellation property: if a != 0, then ab=ac implies b=c.
+// - GCD domains: the existence of greatest common divisor
+// - unique factorization domains: every non-zero element can be uniquely
+//   written as a product of prime elements (for example, 12 = 2*2*3)
+// - Euclidean domains: the fundamental division-with-remainder property (will
+//   explain in section "A Review of Number theory")
+//
+// 3. A Review of Number theory
+// ============================
+//
+// The definition of div and mod comes from the Division Theorem  (aka. the
+// fundamental division-with-remainder property, aka. Euclid's division lemma)
+// described as below:
+//
+// Theorem 3.1 (Division Theorem): for any integers a and b (b != 0), there
+// exist unique integers q and r such that:
+// 1) 0 <= r < |b|
+// 2) a = bq + r
+//
+// if we define a/b := q, and a%b := r, then we can write 2) as:
+//   a = (a/b)b + a%b
+// I will call the above equation "the fundamental division-with-remainder
+// equation" in later contexts.
+//
+// In this section and later sections, I will use \ for real, and / for integer
```

Comment:
Instead of "real" division, I suggest using the term "true division," which is consistent with Python. Real numbers could be divided using floor or ceil or truncation division, too, after all

Reply:
changed